lib/Backends/CPU/CPUFunction.cpp

```diff
@@ -13,28 +13,80 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 #include "CPUFunction.h"
 
+#include "glow/Graph/Context.h"
 #include "glow/Support/Compiler.h"
 #include "glow/Support/Memory.h"
 
 using namespace glow;
 
-CPUFunction::CPUFunction(std::unique_ptr<llvm::orc::GlowJIT> JIT, void *heap)
-    : JIT_(std::move(JIT)), heap_(heap) {}
+CPUFunction::CPUFunction(std::unique_ptr<llvm::orc::GlowJIT> JIT,
+                         runtime::RuntimeBundle runtimeBundle)
+    : JIT_(std::move(JIT)), runtimeBundle_(runtimeBundle) {}
+
+CPUFunction::~CPUFunction() { free(runtimeBundle_.constants); }
 
-CPUFunction::~CPUFunction() { alignedFree(heap_); }
+void CPUFunction::allocateMutableBuffersOnDevice() {
+  baseActivationsAddress_ = (uint8_t *)alignedAlloc(
+      runtimeBundle_.activationsMemSize, TensorAlignment);
+
+  baseMutableWeightVarsAddress_ = (uint8_t *)alignedAlloc(
+      runtimeBundle_.mutableWeightVarsMemSize, TensorAlignment);
+}
+
+void CPUFunction::copyMutablesToDevice(Context &ctx) {
+  // Copy Placeholders into allocated memory.
+  for (auto PH : ctx.pairs()) {
+    auto payload = PH.second->getUnsafePtr();
+    auto symbolInfo =
+        runtimeBundle_.symbolTable.find(std::string(PH.first->getName()));
+    assert(symbolInfo != runtimeBundle_.symbolTable.end() &&
+           "Symbol not found");
+    auto addr = symbolInfo->second.offset;
+    auto numBytes = symbolInfo->second.size;
+    // copy PH to allocated memory.
+    memcpy(baseMutableWeightVarsAddress_ + addr, payload, numBytes);
+  }
+}
+
+void CPUFunction::copyMutablesFromDevice(Context &ctx) {
+  // Copy placeholders from device back into context.
+  for (auto PH : ctx.pairs()) {
+    auto symbolInfo =
+        runtimeBundle_.symbolTable.find(std::string(PH.first->getName()));
+    auto payload = baseMutableWeightVarsAddress_ + symbolInfo->second.offset;
+    auto numBytes = symbolInfo->second.size;
+    auto addr = PH.second->getUnsafePtr();
+    // copy PH from allocated memory.
+    memcpy(addr, payload, numBytes);
+  }
+}
+
+void CPUFunction::freeAllocations() {
+  free(baseMutableWeightVarsAddress_);
+  free(baseActivationsAddress_);
+}
 
 void CPUFunction::execute(Context &ctx) {
+  copyFunctionToDevice();
+  copyConstantsToDevice();
+  allocateMutableBuffersOnDevice();
+  copyMutablesToDevice(ctx);
+
   auto sym = JIT_->findSymbol("jitmain");
   assert(sym && "Unable to JIT the code!");
-  using JitFuncType = void (*)(void);
+  using JitFuncType =
+      void (*)(uint8_t * constantWeightVars, uint8_t * mutableWeightVars,
+               uint8_t * activations);
   auto address = sym.getAddress();
   if (address) {
```

Comment:
I'm curious what is the situation when sym.getAddress() is null? Could it ever happen?

Reply:
> Have you verified that AOT still works :)?

Testing now, it has one core at 100% and has for the last 5 minutes....

Reply:
It is very unlikely to happen, but we can add an assert here just for the case.